File: api_cpp/exponents/PAM_generator.cpp
#include <vector>
#include <cmath>
#include <tuple>

using namespace std;

tuple<vector<double>, vector<double>> PAM_generator(int M) {
    // Probabilities
    vector<double> q(M);
    for (int i = 0; i < M; i++) {
        q[i] = 1.0 / M;
    }

    // Symbols
    vector<double> x(M);
    double aux = -(M - 1);
    for (int i = 0; i < M; i++) {
        x[i] = aux;
        aux += 2;
    }

    // Calculate average energy
    double energy = 0;
    for (size_t i = 0; i < x.size(); i++) {
        energy += q[i] * pow(x[i], 2);
    }

    // Normalize to unit energy
    double norm_factor = sqrt(energy);
    for (auto& symbol : x) {
        symbol /= norm_factor;
    }

    return make_tuple(x, q);
}

File: code/exponents_matlab/PAM_generator.m
function [x,q] = PAM_generator(M)
    % Probabilities
    q = zeros(1,M);
    for i = 1:M
        q(i) = 1/M;
    end

    % Symbols
    x = zeros(1,M);
    aux = -(M - 1);
    for i = 1:M
        x(i) = aux;
        aux = aux + 2;    
    end

    % Average energy
    energy = 0;
    for i = 1:length(x)
        energy = energy + q(i)*abs(x(i))^2;
    end

    % Unit energy
    x = x/sqrt(energy);
    
end
